front/components/agent_builder/instructions/InstructionBlockExtension.tsx

```diff
@@ -0,0 +1,246 @@
+import type { NodeViewProps } from "@tiptap/react";
+import {
+  NodeViewContent,
+  NodeViewWrapper,
+  ReactNodeViewRenderer,
+} from "@tiptap/react";
+import React from "react";
+
+import {
+  createProseMirrorInstructionBlock,
+  INSTRUCTION_BLOCK_REGEX,
+  OPENING_TAG_REGEX,
+  parseInstructionBlockMatches,
+  splitTextAroundBlocks,
+} from "@app/lib/client/assistant_builder/instructionBlockUtils";
+
+export interface InstructionBlockAttributes {
+  type: string;
+}
+
+declare module "@tiptap/core" {
+  interface Commands<ReturnType> {
+    instructionBlock: {
+      setInstructionBlock: (
+        attributes: InstructionBlockAttributes
+      ) => ReturnType;
+      insertInstructionBlock: (type: string) => ReturnType;
+    };
+  }
+}
+
+const InstructionBlockComponent: React.FC<NodeViewProps> = ({ node }) => {
+  const { type } = node.attrs as InstructionBlockAttributes;
+
+  return (
+    <NodeViewWrapper className="my-4">
+      <div className="rounded-lg border border-border bg-muted-background p-4">
+        <div className="mb-3">
+          <span
+            className="text-sm font-semibold text-muted-foreground"
+            contentEditable={false}
+          >
+            {type.toUpperCase()}
+          </span>
+        </div>
+        <div className="min-h-12 rounded border border-border/50 bg-background p-3">
+          <NodeViewContent className="prose prose-sm" />
+        </div>
+      </div>
+    </NodeViewWrapper>
+  );
+};
+
+export const InstructionBlockExtension =
+  Node.create<InstructionBlockAttributes>({
+    name: "instructionBlock",
+    group: "block",
+    content: "block+",
+    defining: true,
+
+    addAttributes() {
+      return {
+        type: {
+          default: "info",
+          parseHTML: (element) =>
+            element.getAttribute("data-instruction-type") || "info",
+          renderHTML: (attributes) => ({
+            "data-instruction-type": attributes.type,
+          }),
+        },
+      };
+    },
+
+    parseHTML() {
+      return [
+        {
+          tag: "div[data-type='instruction-block']",
+        },
+      ];
+    },
+
+    renderHTML({ HTMLAttributes }) {
+      return [
+        "div",
+        mergeAttributes(HTMLAttributes, {
+          "data-type": "instruction-block",
+        }),
+        0,
+      ];
+    },
+
+    addCommands() {
+      return {
+        setInstructionBlock:
+          (attributes) =>
+          ({ commands }) =>
+            commands.setNode(this.name, attributes),
+
+        insertInstructionBlock:
+          (type) =>
+          ({ commands }) =>
+            commands.insertContent({
+              type: this.name,
+              attrs: { type },
+              content: [{ type: "paragraph" }],
+            }),
+      };
+    },
+
+    addNodeView() {
+      return ReactNodeViewRenderer(InstructionBlockComponent);
+    },
+
+    addInputRules() {
+      return [
+        new InputRule({
+          find: OPENING_TAG_REGEX,
```

Comment:
Do you know why this doesn't work if you start with empty <>? 🤔

https://github.com/user-attachments/assets/1cefddf3-b650-48d1-b102-059137698942

Reply:
No clue.. but will dig if we want to keep it 👍🏼